tagmanager/src/tm_workspace.c

```diff
@@ -823,514 +728,333 @@ fill_find_tags_array (GPtrArray *dst, const GPtrArray *src,
+	langType lang)
+{
+	const gchar *sep = scope_sep(lang);
+	gboolean ret = FALSE;
+	gchar **comps;
+	guint len;
+
+	/* method scope is in the form ...::class_name::method_name */
+	comps = g_strsplit (method_scope, sep, 0);
+	len = g_strv_length(comps);
+	if (len > 1)
 	{
-		const GPtrArray *tags2;
-		guint got = 0;
-		TMTagType types = (tm_tag_class_t
-						   tm_tag_struct_t | tm_tag_typedef_t |
-						   tm_tag_union_t | tm_tag_enum_t);
+		gchar *method, *inner_scope, *cls, *cls_scope;
+		GPtrArray *tags;
+
+		/* get method/member scope */
+		method = comps[len - 1];
+		comps[len - 1] = NULL;
+		inner_scope = g_strjoinv(sep, comps);
+		comps[len - 1] = method;
+
+		/* get class scope */
+		cls = comps[len - 2];
+		comps[len - 2] = NULL;
+		cls_scope = g_strjoinv(sep, comps);
+		comps[len - 2] = cls;
+
+		/* check whether the class exists */
+		tags = tm_workspace_find(cls, cls_scope, tm_tag_class_t | tm_tag_struct_t, NULL, FALSE, lang);
+		ret = tags->len > 0;
+
+		/* check whether member inside the class */
+		if (ret)
+			ret = g_strcmp0(member_scope, inner_scope) == 0;
+
+		g_ptr_array_free(tags, TRUE);
+		g_free(cls_scope);
+		g_free(inner_scope);
+	}
 
-		if (file_tags)
-		{
-			g_ptr_array_set_size (tags, 0);
-			got = fill_find_tags_array (tags, file_tags,
-										  new_name, NULL, types, FALSE, -1, FALSE);
-		}
+	g_strfreev(comps);
+	return ret;
+}
 
 
-		if (got)
-		{
-			tags2 = tags;
-		}
-		else
-		{
-			TMTagAttrType attrs[] = {
-				tm_tag_attr_name_t, tm_tag_attr_type_t,
-				tm_tag_attr_none_t
-			};
-			tags2 = tm_workspace_find (new_name, types, attrs, FALSE, -1);
-		}
+/* For an array of variable/type tags, find members inside the types */
+static GPtrArray *
+find_scope_members_all(GPtrArray *tags, GPtrArray *searched_array, langType lang,
+	gboolean member, const gchar *current_scope)
+{
+	GPtrArray *member_tags = NULL;
+	guint i;
 
-		if ((tags2) && (tags2->len == 1) && (tag = TM_TAG (tags2->pdata[0])))
+	/* there may be several variables/types with the same name - try each of them until
+	 * we find something */
+	for (i = 0; i < tags->len; i++)
+	{
+		TMTag *tag = TM_TAG(tags->pdata[i]);
+		TMTagType types = (tm_tag_class_t | /* tm_tag_namespace_t | tm_tag_package_t | */
+						   tm_tag_struct_t | tm_tag_union_t |
+						   tm_tag_enum_t | tm_tag_typedef_t);
+
+		if (tag->type & types)  /* type: namespace search */
+			member_tags = find_scope_members(searched_array, tag->name, lang, TRUE);
+		else if (tag->var_type)  /* variable: scope search */
 		{
-			if (tag->type == tm_tag_typedef_t && tag->var_type
-				&& tag->var_type[0] != '\0')
-			{
-				new_name = tag->var_type;
-				continue;
-			}
-			filename = (tag->file ?
-						tag->file->short_name : NULL);
-			if (tag->scope && tag->scope[0] != '\0')
+			if (!(tag->type & (tm_tag_field_t | tm_tag_member_t)) || member ||
+				member_at_method_scope(current_scope, tag->scope, lang))
```

Comment:
an explanation of this check might be useful for the poor guy trying to understand why it's like that in the future ^^